text/0401-coercions.md

```diff
@@ -319,20 +319,23 @@ descriptions are equivalent.
 Casting is indicated by the `as` keyword. A cast `e as U` is valid if one of the
 following holds:
 
-* `e` has type `T` and `T` coerces to `U`;
-
-* `e` has type `*T` and `U` is `*U_0` (i.e., between any raw pointers);
-
-* `e` has type `*T` and `U` is `uint` , or vice versa;
-
-* `e` has type `T` and `T` and `U` are any numeric types;
-
-* `e` is a C-like enum and `U` is any integer type, `bool`;
-
-* `e` has type `T` and `T == u8` and `U == char`;
-
-* `e` has type `T` and `T == &[V, ..n]` or `T == &V` and `U == *const V`, and
-  similarly for the mutable variants to either `*const V` or `*mut V`.
+ * `e` has type `T` and `T` coerces to `U`; *coercion-cast*
+ * `e` has type `*T`, `U` is `*U_0`, and either `U_0: Sized` or
+    unsize_kind(`T`) = unsize_kind(`U_0`); *ptr-ptr-cast*
+ * `e` has type `*T` and `U` is a numeric type, while `T: Sized`; *ptr-addr-cast*
+ * `e` is an integer and `U` is `*U_0`, while `U_0: Sized`; *addr-ptr-cast*
+ * `e` has type `T` and `T` and `U` are any numeric types; *numeric-cast*
+ * `e` is a C-like enum and `U` is an integer type; *enum-cast*
+ * `e` has type `bool` or `char` and `U` is an integer; *prim-int-cast*
+ * `e` has type `u8` and `U` is `char`; *u8-char-cast*
+ * `e` has type `&[T; n]` and `U` is `*const T`; *array-ptr-cast*
```

Comment:
@nrc 

I can't think of a reason to exclude `mut` here. After all, you can certainly do `&mut foo[0] as *T`. Regarding 0, I think the answer is probably "an undefined value" (not UB) for the time being, but in general we actually are more limiting than you might think here. (Specifically, we have be careful because of zero-sized types; we used to use `null` more often, but often now use `heap::EMPTY`, because of interactions with the null-pointer-optimizations and so forth. A specific example is escaping me at this moment.)